### DISCUSSION TITLE
How do you convert a SQL formula to Javascript API?

### DISCUSSION THREAD

**Discussion by belsandre** [supabase/supabase#3570] [Questions]
I want to use this SQL formula to find the closest 20 locations in my Supabase database to any arbitrary location:

  SELECT 
  id, 
  (
     3959 *
     acos(cos(radians(37)) * 
     cos(radians(lat)) * 
     cos(radians(lng) - 
     radians(-122)) + 
     sin(radians(37)) * 
     sin(radians(lat )))
  ) AS distance 
  FROM markers 
  HAVING distance < 28 
  ORDER BY distance LIMIT 0, 20;

How do I convert this to Supabase's Javascript API? I don't see a way to add formulas or substitutes for "AS" and "HAVING" operators

Thanks!

**Comment by VictorPeralta** [comment]
You can create a database function for this and then call it throug the JS client.

**Comment by silentworks** [comment]
If these values are set values and don't need them to be dynamic then you can create a Postgres view (you can call a view using the same `supabase.from('view_name')`, otherwise, if the values are dynamic then use a Postgres function and call it via `supabase.rpc('function_name')`.

https://supabase.io/docs/reference/javascript/rpc

**Reply by belsandre** [reply]
It's dynamic, so I think I need to create a Postgres function. However I've tried a bunch of things and it keeps saying "Success. No rows returned." I've installed earthdistance extension, so I'm using point() from that extension. How can I fix this?

```
drop function if exists find_closest_clinics();

create or replace function find_closest_clinics(from_lat float, from_lon float)
returns TABLE(id int, name text, city text, state text, distance float) 
as $$
  SELECT 
    id, name, city, state, 
    (point(from_lon, from_lat)<@>point(lon, lat)) as distance
  FROM clinics 
  GROUP BY id 
  HAVING (point(from_lon, from_lat)<@>point(lon, lat)) < 1000
  ORDER BY distance LIMIT 10;
$$ language sql stable;
```

Thanks!
